### ISSUE TITLE
BUG: Error on Lint SQL file Contains Custom Created-macros in DBT

### ISSUE THREAD

**Issue by tvlk-karl** [sqlfluff/sqlfluff#1122]
the error showed up when I try to do `sqlfluff lint` on sql files which contains custom created-macros using dbt and bigquery
```
WARNING    Unable to lint /app/dbt/models/fpp/dummy_l3/test4.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 1415, in lint_path
    linted_file = self.lint_string(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 1198, in lint_string
    parsed = self.parse_string(in_str=in_str, fname=fname, config=config)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 904, in parse_string
    tokens, lex_vs = lexer.lex(templated_file)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/parser/lexer.py", line 319, in lex
    segments: Tuple[RawSegment, ...] = self.elements_to_segments(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/parser/lexer.py", line 343, in elements_to_segments
    source_slice = templated_file.templated_slice_to_source_slice(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/templaters/base.py", line 287, in templated_slice_to_source_slice
    start_slices = [self.sliced_file[ts_start_sf_start]]
IndexError: list index out of range
```
it starts to show error when hit file `test4.sql`

## Expected Behaviour
it should be no error, as dbt can run it

## Steps to Reproduce
```sql
# file: test.sql
{{
  config(
    materialized="table"
  )
}}

SELECT 'test-fpp-dbt' AS col, CAST('2021-01-01' AS DATE) AS dt, 15 AS num
UNION ALL
SELECT 'test-fpp-dbt' AS col, CAST('2021-02-01' AS DATE) AS dt, 30 AS num
UNION ALL
SELECT 'test-fpp-dbt' AS col, CAST('2021-01-01' AS DATE) AS dt, 25 AS num
```
```sql
# file: test4.sql
{{
    config(
        materialized="table",
    )
}}

WITH tt AS
(
    SELECT *
    FROM {{ ref('test') }}
)

{{ get_latest_row_only('tt', ['col'], 'num') }}

SELECT *
FROM tt_latest
```
```sql
# name: get_latest_row_only.sql
-- Macro to add a row number when partitioning by the list of partition_fields,
-- ordering by order_field descending and only getting row number = 1 (and
-- discarding the row number field)

{% macro get_latest_row_only(from_table, partition_fields, order_field) %}
{% set intermediate_table = from_table + '_rowed' %}
{% set target_table = from_table + '_latest' %}
, {{ intermediate_table }} AS (
  SELECT
    *
    , ROW_NUMBER() OVER (
        PARTITION BY
          {% for partition_field in partition_fields %}
            {{ partition_field }} {% if not loop.last %}, {% endif %}
          {% endfor %}
        ORDER BY {{ order_field }} DESC
      ) AS row_number
  FROM
    {{ from_table }}
)

, {{ target_table }} AS (
  SELECT
    * EXCEPT(row_number)
  FROM
    {{ intermediate_table }}
  WHERE
    row_number = 1
)
{% endmacro %}
```

## Version
Include the output of `sqlfluff --version` along with your Python version
`sqlfluff: 0.5.6`
`python: 3.8.8`

## Configuration
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = bigquery
templater = dbt

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path = macros
```

**Comment by hugohjerten** [comment]
I'm experiencing the exact same problem when using custom macros when using dbt and bigquery.

**Comment by Startouf** [comment]
Same problem using sqlfluff 0.6.0

**Comment by barrywhart** [comment]
I tested this today on the latest code in `main` (commit `826798198aa663739ad81ce7b1b3e4192b4d4ab9`). Both `test.sql` and `test4.sql` were successfully linted. Output below.
```
(sqlfluff-3.7.9) ➜  my-dbt sqlfluff lint models/
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
Looking for: /Users/bhart/dev/my-dbt/models/example/test.sql
Success!
== [models/example/test.sql] FAIL
L:   7 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
L:   9 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
L:  11 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
Looking for: /Users/bhart/dev/my-dbt/models/example/test4.sql
Success!
== [models/example/test4.sql] FAIL
L:   7 | P:   9 | L023 | Single whitespace expected after AS in WITH clause.
L:  15 | P:   1 | L044 | Query produces an unknown number of result columns.
All Finished 📜 🎉!
```

Closing issue.
